### PR TITLE
subsys/ztest: Make 1cpu tests run on CPU 0 specifically

### PR DESCRIPTION
**Pull request by andyross**
Some hardware has "interesting" configuration like asymmetric default
interrupt masking (the intel_adsp devices in particular, but x86's
IO-APIC driver has tripped over this in the past too) that needs
special treatment if you want to run something on "core 1"
specifically, and 1cpu test cases pretty much by definition are going
to have been written without SMP details in mind.  Switch the logic
around a tiny bit such that these test cases always run on CPU ID zero
explicitly.

Actually in practice this was ALMOST guaranteed to be true already,
because test setup and happens serially, having been started on the
main thread, which starts on CPU 0 by definition.  Then the test
teardown aborts all the spawned threads that might have been running
on CPUs 1+, so those reach idle, and the next test case starts
syncronously on the same thread (and thus CPU) where it started.  But
nonetheless that wasn't actually enforced, and we've found at least
one simulation environment where timing conspires to break things.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

### CODE REVIEW

**Comment by andyross** [comment]
Adding a reasonably obvious feature to ztest to help chase a great hypothesis from @enjiamai

**Comment by enjiamai** [comment]
Hi @andyross , Sorry for a wrong hypothesis and misleading you. I will do more analysis on this.

**Review by dcpleung — APPROVED** [review]
Wonder if we should just enable CPU mask on ztest with SMP.

**Comment by andyross** [comment]
> Wonder if we should just enable CPU mask on ztest with SMP.

That would indeed be easier, but then you can't get coverage of !CPU_MASK though, thus making it an essentially mandatory feature.

Really this isn't that complicated.  Or... it's not any more complicated than the original hack anyway.

**Review by ceolin — APPROVED** [review]

File: subsys/testsuite/ztest/src/ztest.c [hunk 1]
```diff
@@ -90,33 +90,30 @@ static int cleanup_test(struct unit_test *test)
...
-#ifdef CONFIG_SMP
-#define NUM_CPUHOLD (CONFIG_MP_NUM_CPUS - 1)
-#else
-#define NUM_CPUHOLD 0
-#endif
 #define CPUHOLD_STACK_SZ (512 + CONFIG_TEST_EXTRA_STACKSIZE)
 
-static struct k_thread cpuhold_threads[NUM_CPUHOLD];
-K_KERNEL_STACK_ARRAY_DEFINE(cpuhold_stacks, NUM_CPUHOLD, CPUHOLD_STACK_SZ);
+#if defined(SMP) && (CONFIG_MP_NUM_CPUS > 1)
```
Comment (ceolin) [inline subsys/testsuite/ztest/src/ztest.c:95]: s/SMP/CONFIG_SMP/

File: subsys/testsuite/ztest/src/ztest.c [hunk 2]
```diff
@@ -90,33 +90,30 @@ static int cleanup_test(struct unit_test *test)
...
  * may need to control that configuration themselves.  We do this at
  * the lowest level, but locking interrupts directly and spinning.
  */
-static void cpu_hold(void *arg1, void *arg2, void *arg3)
+static inline void cpu_hold(void *arg1, void *arg2, void *arg3)
 {
 	ARG_UNUSED(arg1);
 	ARG_UNUSED(arg2);
 	ARG_UNUSED(arg3);
+#if defined(SMP) && (CONFIG_MP_NUM_CPUS > 1)
```
Comment (ceolin) [inline subsys/testsuite/ztest/src/ztest.c:113]: s/SMP/CONFIG_SMP/

File: subsys/testsuite/ztest/src/ztest.c [hunk 3]
```diff
@@ -167,21 +183,23 @@ void z_impl_z_test_1cpu_start(void)
...
+
+	/* Sleep, waiting to be woken up on cpu0 */
+	k_sem_take(&cpuhold_sem, K_FOREVER);
+	__ASSERT(arch_curr_cpu()->id == 0, "1cpu case running on wrong cpu");
+#endif
 }
 
 void z_impl_z_test_1cpu_stop(void)
 {
+#if defined(SMP) && (CONFIG_MP_NUM_CPUS > 1)
```
Comment (ceolin) [inline subsys/testsuite/ztest/src/ztest.c:196]: s/SMP/CONFIG_SMP/

**Review by ceolin — CHANGES_REQUESTED** [review]
The build condition is wrong and the code is never executed, so cpus are not hold.

I've tested with the change and it works great !

**Comment by andyross** [comment]
Whoops, yeah.  That snuck in at the end, I changed the #if logic so the threads wouldn't be defined if not needed and clearly that snuck in.

**Review by ceolin — APPROVED** [review]

**Review by enjiamai — CHANGES_REQUESTED** [review]
Excuse me, @andyross , I would like to block this first, I made a full testing on it, the result is better but not address all the problem.  Let me provide you more details. Thanks!

**Comment by ceolin** [comment]
Is there any reason to not merget it ?

**Comment by enjiamai** [comment]
Hi @ceolin @andyross , I think we don't need this because the original one works well now and it is a real SMP scenario. But this might be possibly needed in the future if we meet the scenario that "main core" cannot use for spinning around. How do you think adding a CONFIG_MAIN_CPU_NO_SPIN to enable/disable this feature is a good idea for this?

**Comment by andyross** [comment]
This no longer fixes anything.  It's somewhat reasonable protection against future systems with the kind of CPU asymmetry we were worried about, and matches the broad intent of the API (to emulate a "one cpu" system as closely as possible for the benefit of tests that aren't MP-aware).  I'm happy if this goes in or gets dropped, no preference.  Though it looks like it does need a rebase.

**Comment by ceolin** [comment]
> This no longer fixes anything.  It's somewhat reasonable protection against future systems with the kind of CPU asymmetry we were worried about, and matches the broad intent of the API (to emulate a "one cpu" system as closely as possible for the benefit of tests that aren't MP-aware).  I'm happy if this goes in or gets dropped, no preference.  Though it looks like it does need a rebase. 

Exactly ... +1 to merge it.

**Review by enjiamai — APPROVED** [review]

**Comment by andyross** [comment]
Huh... so I pushed a new tree to https://github.com/andyross/zephyr/commits/1cpu0 to fix the rebase collision, but it's not being picked up here.  Is this a github glitch or is there something I have to do to kick it?

**Comment by nashif** [comment]
needs rebase

**Comment by andyross** [comment]
Still stuck, despite source branch being updated.  Try closing/reopening...

**Comment by andyross** [comment]
And... now it won't let me reopen because "The 1cpu0 branch was force-pushed or recreated".  Well, yeah.  Why is that a problem?

Did we break something with the github settings?

**Comment by andyross** [comment]
Yeah, this is borked.  Opened a new PR at #43547

**Pull request closed without merging**